File: Back-End/src/main/java/model/Professor.java
package model;

import org.bson.Document;
import org.json.JSONArray;
import org.json.JSONObject;
import util.ExceptionHandler;

import java.util.ArrayList;
import java.util.List;

public class Professor extends Usuario {
    private int idProfessor;
    private List<Integer> disciplinasLecionadas;
    public Professor(String nome,String nomeUsuario,String senha,String tipo,String email,
                     int idProfessor,List<Integer> disciplinasLecionadas){
        super(nome,nomeUsuario,senha,tipo,email);
        this.idProfessor=idProfessor;
        this.disciplinasLecionadas=disciplinasLecionadas;
    }
    public int getIdProfessor(){
        return this.idProfessor;
    }
    public List<Integer> getDisciplinasLecionadas(){
        return this.disciplinasLecionadas;
    }

    public Document paraJson(){
        try {


            JSONObject professorJson = new JSONObject();

            // Dados do usuario
            professorJson.put("nome", super.getNome());
            professorJson.put("nome_usuario", super.getNomeUsuario());
            professorJson.put("senha", super.getSenha());
            professorJson.put("email", super.getEmail());
            professorJson.put("tipo", "professor");


            //dados especificos do professor devem ser salvos em outro json
            JSONObject dadosProfessor = new JSONObject();
            dadosProfessor.put("id_professor", this.idProfessor);

            // tranformando a lista de disciplinasLecionadas para um array para amarzenar
            JSONArray arrayDisciplinas = new JSONArray(this.disciplinasLecionadas);

            // adicionando o array de disciplinas no campo
            dadosProfessor.put("disciplinaLecionadas", arrayDisciplinas);


            professorJson.put("dados_professor", dadosProfessor);


            return Document.parse(professorJson.toString());
        } catch (Exception e) {
            ExceptionHandler eh = new ExceptionHandler(e);
            eh.printExeption();
            return null;
        }
    }





    public static Professor deJson(Document jsonDoc) {
        try{
            // Extrair dados do usuário
            String nome = jsonDoc.getString("nome");
            String nomeUsuario = jsonDoc.getString("nome_usuario");
            String senha = jsonDoc.getString("senha");
            String email = jsonDoc.getString("email");
            String tipo = jsonDoc.getString("tipo");

            // Extrair dados específicos do professor
            Document dadosProfessor= jsonDoc.get("dados_professor", Document.class);

            int idProfessor = dadosProfessor.getInteger("id_professor");

            // pegando o array de materias lecionadas e transformando em uma List
            List<Integer> disciplinasLecionadas = dadosProfessor.getList("disciplinaLecionadas", Integer.class);

            // garatindo que o array não vai estar vazio e se estiver volta uma lista vazia
            if (disciplinasLecionadas == null) {
                disciplinasLecionadas = new ArrayList<>();
            }



            // Criar e retornar o objeto Aluno
            return new Professor(
                    nome, nomeUsuario, senha, tipo, email,
                    idProfessor, disciplinasLecionadas
            );
        }catch (Exception e) {
            ExceptionHandler eh = new ExceptionHandler(e);
            eh.printExeption();
            return null;
        }
    }
    @Override
    public String toString() {
        return "Professor{" +
                "idProfessor=" + idProfessor +
                ", disciplinasLecionadas=" + disciplinasLecionadas +
                '}';
    }
}

File: Back-End/src/main/JavaScript/code/professor/Professor.js
import Usuario from "./Usuario.js";

class Professor extends Usuario {
    constructor(nome, nomeUsuario, senha, tipo, email, idProfessor, disciplinasLecionadas, id) {
        super(nome, nomeUsuario, senha, tipo, email, id);
        this.idProfessor = idProfessor;
        this.disciplinasLecionadas = disciplinasLecionadas;
    }

    getIdProfessor() {
        return this.idProfessor;
    }

    getDisciplinasLecionadas() {
        return this.disciplinasLecionadas;
    }

    setSenha(senha) {this.senha = senha;}


    paraJson() {
        try {
            const professorJson = {
                nome: super.getNome(),
                nome_usuario: super.getNomeUsuario(),
                senha: super.getSenha(),
                email: super.getEmail(),
                tipo: "professor",
                dados_professor: {
                    id_professor: this.idProfessor,
                    disciplinaLecionadas: this.disciplinasLecionadas
                }
            };
            return professorJson;
        } catch (e) {
            console.log(e);
            return null;
        }
    }

    static deJson(jsonDoc) {
        try {
            const id = jsonDoc._id ? (jsonDoc._id.$oid || jsonDoc._id) : null;
            const nome = jsonDoc.nome;
            const nomeUsuario = jsonDoc.nome_usuario;
            const senha = jsonDoc.senha;
            const email = jsonDoc.email;
            const tipo = jsonDoc.tipo;

            const dadosProfessor = jsonDoc.dados_professor;
            const idProfessor = dadosProfessor.id_professor;
            
            let disciplinasLecionadas = dadosProfessor.disciplinaLecionadas;
            if (!disciplinasLecionadas) {
                disciplinasLecionadas = [];
            }

            return new Professor(
                nome, nomeUsuario, senha, tipo, email,
                idProfessor, disciplinasLecionadas, id
            );
        } catch (e) {
            console.log(e);
            return null;
        }
    }

    toString() {
        return JSON.stringify({
            id: super.getId(),
            nome: super.getNome(),
            nomeUsuario: super.getNomeUsuario(),
            email: super.getEmail(),
            tipo: super.getTipo(),
            idProfessor: this.idProfessor,
            disciplinasLecionadas: this.disciplinasLecionadas || []
        }, null, 2);
    }
}

export default Professor;